Fix Git dependency version detection and improve output

neilberkman opened the pull request:
Closes #6

## Summary

This PR fixes a critical bug where Git dependencies were incorrectly showing as changed on every run of the deps_changelog task. The fix properly extracts commit hashes from the lock field and adds informative output messages.

## Changes

### Bug Fix
- Fixed Git dependency version extraction: Now correctly uses commit hashes from `Mix.Dep.opts[:lock]` instead of semantic versions from the status field
- Proper version comparison: Git dependencies are now compared by their actual commit hashes, eliminating false positives

### Improvements
- Repository info display: Git dependencies now show as `owner/repo@commit` format (e.g., `serpent213/deps_changelog@b1a69cf8`)
- Informative output messages: 
  - Shows detected dependency changes with old → new versions
  - Reports "No dependency changes detected" when appropriate
  - Uses debug level for verbose progress messages
- Better handling of edge cases:
  - Transitions between Git and Hex dependencies
  - GitHub fork switches
  - Atom sources (e.g., `:heroicons`)

### Testing
- Added comprehensive test coverage for:
  - Git dependency changes
  - Hex to Git transitions
  - Fork switching scenarios
  - Mixed dependency updates

## Before/After

**Before:**
```
# Git dependencies always showed as changed, even when unchanged
```

**After:**
```
Detected dependency changes:
  • doctest_formatter: neilberkman/doctest_formatter@bb0c7b8d → 0.4.1
  • deps_changelog: serpent213/deps_changelog@b1a69cf8 → neilberkman/deps_changelog@c3300504
```

## Impact

This fix makes the changelog accurate and useful for projects using Git dependencies, eliminating the noise from false positives that made the tool less valuable.

neilberkman committed:
Fix Git dependency version detection and improve output clarity

This fixes a critical bug where unchanged Git dependencies were incorrectly
reported as changed due to comparing semantic versions from status field
instead of commit hashes from the lock.

Changes:
- Use commit hash from lock for Git dependency comparisons
- Display semantic versions for Hex packages (e.g., "1.7.0")
- Display repository info for Git deps (e.g., "owner/repo@abc123de")
- Detect repository changes when switching forks
- Add comprehensive test coverage for all dependency scenarios
- Improve output messages for better clarity:
  * Show dependency/changelog counts in snapshot
  * List detected changes before processing
  * Clarify why changelog wasn't updated
  * Use debug level for verbose progress messages

The fix ensures Git dependencies only show as changed when their commit
hash actually changes, eliminating false positives in changelog tracking.

neilberkman committed:
Remove debug statements - Mix.Shell.IO.debug is not available

serpent213 commented:
Aah, I made a mess...!  😅

Looks great, would you please rebase against `develop` and `mix format`?

neilberkman committed:
Add guard clause for safer Git dependency pattern matching

As suggested by code review, ensure both url and commit are binary strings
when pattern matching Git dependencies to prevent potential runtime errors
if the lock tuple structure changes in future Mix versions.
